docs-2.0/1.introduction/3.nebula-graph-architecture/4.storage-service.md

```diff
@@ -1,11 +1,261 @@
+
+Read and write: For every writing request of the clients, the Leader will initiate a Raft-wal and synchronize it with the Followers. Only after over half replicas have received the Raft-wal will it return to the clients successfully. For every reading request of the clients, it will get to the Leader directly, while Followers will not be involved.
+
+Failure: If the system has only one replica, Leader will be itself. If failure happens, the system will be completely unavailable. If the system has three replicas, one of them will be Leader and the rest will be Followers. If the Leader fails, the rest two can still vote for a new Leader (and a Follower), and the system is still available. But if any of the two Followers fails again, the system will be completely unavailable due to inadequate voters.
+
+!!! Note
+
+    Raft and HDFS have different modes of duplication. Raft is based on a majority vote, so the number of replicas cannot be even.
+
+Listener: As is a special role in Raft, it cannot vote or keep data consistency. In Nebula Graph, it reads Raft-wal from Leader and synchronizes it to ElasticSearch cluster.
+
+### Multi Group Raft
+
+Storage service supports a distributed cluster architecture, so Nebula Graph implements Multi Group Raft according to Raft requirement. Each Raft group stores all the replicas of each partition. One replica is the leader, while others are followers. In this way, Nebula Graph achieves strong consistency and high availability. The functions of Raft are as follows.
+
+Nebula Graph uses Multi Group Raft to improve performance when there are many partitions because Raft-wal cannot be NULL. When there are too many partitions, costs will increase, such as storage information in Raft group, WAL files, or batch operation in low load.
+
+There are two key points to implement the Multi Raft Group:
+
+- To share transport layer
+
+   Each Raft Group sends messages to its corresponding peers. So if the transport layer cannot be shared, the connection costs will be very high.
+
+- To share thread pool
+
+  Raft Groups share the same thread pool to prevent starting too many threads and a high context switch cost.
+
+### Batch
+
+For each partition, it is necessary to do a batch to improve throughput when writing the WAL serially. As Nebula Graph uses WAL to implement some special functions, batches need to be grouped, which is a feature of Nebula Graph.
+
+For example, lock-free CAS operations will execute after all the previous WALs are committed. So for a batch, if there are several WALs in CAS type, we need to divide this batch into several smaller groups and make sure they are committed serially.
+
+### Listener
+
+Listener is designed for **storage expansion**. It takes a long time for the newly added machines to synchronize data during expansion. If synchronizing as a follower, the availability of the entire cluster will decrease.
```

Comment:
storage horizontal scaling. 

Listener is designed for **storage horizontal scaling. **. It takes a long time for the newly added machines to be synchronized with data. Therefore these manchine can't join the group followers, otherwise, the availability of the entire cluster will decrease.